File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = (int)height.size();
        int l = 0;
        int r = n-1;
        int leftMax = 0;
        int rightMax = 0;
        int ans = 0;
        while (l < r){
            leftMax = max(leftMax,height[l]);
            rightMax = max(rightMax,height[r]);
            if (leftMax <= rightMax){
                ans += leftMax - height[l];
                l++;
            } else {
                ans += rightMax - height[r];
                r--;
            }
        }
        return ans;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: list[int]) -> int:
        l = 0
        r = len(height)-1
        lMax = height[l]
        rMax = height[r]
        ans  = 0
        while l < r:
            lHeight = height[l]
            rHeight = height[r]
            lMax = max(lMax,lHeight)
            rMax = max(rMax,rHeight)
            if lMax < rMax:
                ans += lMax - lHeight
                l += 1
            else:
                ans += rMax - rHeight
                r -= 1
        return ans